### PR TITLE
8317819: Scope should reflect lifetime of underying resource (mainline)

### PR DESCRIPTION
**Pull request by mcimadamore**
This PR is based on a port of: openjdk/panama-foreign#898

Some additional changes are contained in this PR:

* instead of using a different "always alive" scope for zero-length memory segments, we now use the global scope
* the javadoc for `MemorySegment::ofArray`, `MemorySegment::ofBuffer` and `SymbolLookup::loaderLookup` is tweaked to say that the scope associated to the returned segments is an automatic scope
* the javadoc for `MemorySegment.Scope` defines the concepts of global and automatic scope, which are then referenced in the `Arena` javadoc.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8317820](https://bugs.openjdk.org/browse/JDK-8317820) to be approved

### Issues
 * [JDK-8317819](https://bugs.openjdk.org/browse/JDK-8317819): Scope should reflect lifetime of underying resource (mainline) (**Enhancement** - P2)
 * [JDK-8317820](https://bugs.openjdk.org/browse/JDK-8317820): Scope should reflect lifetime of underying resource (mainline) (**CSR**)

### Reviewers
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - **Reviewer**) ⚠️ Review applies to [81e6560d](https://git.openjdk.org/jdk/pull/16183/files/81e6560dace47201ac80ef42ff8f85d253c6e195)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16183/head:pull/16183` \
`$ git checkout pull/16183`

Update a local copy of the PR: \
`$ git checkout pull/16183` \
`$ git pull https://git.openjdk.org/jdk.git pull/16183/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16183`

View PR using the GUI difftool: \
`$ git pr show -t 16183`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16183.diff">https://git.openjdk.org/jdk/pull/16183.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16183#issuecomment-1761622559)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/foreign/Arena.java [hunk 1]
```diff
@@ -43,17 +43,19 @@
  * to obtain native segments.
  * <p>
  * The simplest arena is the {@linkplain Arena#global() global arena}. The global arena
- * features an <em>unbounded lifetime</em>. As such, native segments allocated with the global arena are always
- * accessible and their backing regions of memory are never deallocated. Moreover, memory segments allocated with the
- * global arena can be {@linkplain MemorySegment#isAccessibleBy(Thread) accessed} from any thread.
+ * features an <em>unbounded lifetime</em>. The scope of the global arena is the global scope.
+ * As such, native segments allocated with the global arena are associated are always accessible and their backing regions
```
Comment (JornVernee) [inline src/java.base/share/classes/java/lang/foreign/Arena.java:47]: Typo?
```suggestion
 * As such, native segments allocated with the global arena are always accessible and their backing regions
```

File: src/java.base/share/classes/java/lang/foreign/MemoryLayout.java [hunk 2]
```diff
@@ -460,7 +460,7 @@ default MethodHandle byteOffsetHandle(PathElement... elements) {
      * </ul>
      * <p>
      * If the selected layout is an {@linkplain AddressLayout address layout}, calling {@link VarHandle#get(Object...)}
-     * on the returned var handle will return a new memory segment. The segment is associated with a fresh scope that is
+     * on the returned var handle will return a new memory segment. The segment is associated with a scope that is
```
Comment (JornVernee) [inline src/java.base/share/classes/java/lang/foreign/MemoryLayout.java:463]: Should this say 'global scope' as well?

File: src/java.base/share/classes/java/lang/foreign/MemorySegment.java [hunk 3]
```diff
@@ -2365,11 +2363,38 @@ static long mismatch(MemorySegment srcSegment, long srcFromOffset, long srcToOff
 
     /**
      * A scope models the <em>lifetime</em> of all the memory segments associated with it. That is, a memory segment
-     * cannot be accessed if its associated scope is not {@linkplain #isAlive() alive}. A new scope is typically
-     * obtained indirectly, by creating a new {@linkplain Arena arena}.
+     * cannot be accessed if its associated scope is not {@linkplain #isAlive() alive}. Scope instances can be compared
+     * for equality. That is, two scopes are considered {@linkplain #equals(Object)} if they denote the same lifetime.
```
Comment (JornVernee) [inline src/java.base/share/classes/java/lang/foreign/MemorySegment.java:2367]: missing plain text in the link tag

**Review by JornVernee — APPROVED** [review]

File: src/java.base/share/classes/java/lang/foreign/MemoryLayout.java [hunk 4]
```diff
@@ -481,7 +481,7 @@ default MethodHandle byteOffsetHandle(PathElement... elements) {
      * {@snippet lang = "java":
      * VarHandle baseHandle = this.varHandle(P);
      * MemoryLayout target = ((AddressLayout)this.select(P)).targetLayout().get();
-     * VarHandle targetHandle = target.varHandle(P');
+     * VarHandle targetHandle = target.varHandle(P);
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/MemoryLayout.java:484]: I've seen this typo in the example, and decided to fix it (even if unrelated)

**Review by JornVernee — APPROVED** [review]

**Pull request closed without merging**